### ISSUE TITLE
Toxiproxy and Oracle SCAN

### ISSUE THREAD

**Issue by cjmckenna** [Shopify/toxiproxy#177]
Having a challenge with proxying connections from my application to Oracle databases utilizing the SCAN technology.  What we see is that after the initial connect using the url that has the SCAN address, we subsequently get connected to any one of the actual DB server addresses and.  We can proxy the original "connect" OK but the queries after can take a route of any one of the available servers in the RAC

How do we deal with these dynamic redirects in toxiproxy.

**Comment by jpittis** [comment]
Hey there! Thanks for opening an issue. If you could provide a little more information about what kind of requests are being made and how you're configuring toxiproxy, this may help us understand what's going on.

So I assume you have toxiproxy configured something like this:

```
listen: 127.0.0.1:7878, upstream: <SCAN>:<oracle-port>
```

When the initial connect happens, toxiproxy resolves the SCAN into an IP for a node in the cluster and opens a TCP connection to this node. Future traffic over this connection will go to the same node but if you try to open another TCP connection to toxiproxy, there's a chance that the SCAN will be resolved to a different IP and therefore a different node?

It seems like if you want to use a "multicast" host (the SCAN in this case), you can't expect each connection to toxiproxy to resolve to the same host. If you need all connections to the proxy to go to the same host, you'll need to configure a domain that will resolve to a single host or hardcode a real IP address.

I hope this helps.

**Comment by jpittis** [comment]
Closing due to inactivity and most likely font fix.

**Comment by hberg32** [comment]
I'd like to reopen this issue as I believe I am experiencing the same thing.  I'm not entirely sure if my organization is using Oracle SCAN (Single Client Access Name) or if we just have our DNS set to resolve a single hostname to a URL.  However, let's assume the hostname is oracle-db.mydomain.com and that if I do an nslookup of oracle-db.mydomain.com I see that it resolves to 3 IPs 1.2.3.4, 1.2.3.5, and 1.2.3.6.  If I set up my java application to do a JDBC connection to oracle-db.mydomain.com or to one of the specific IPs then everything is ok.  However, if I set up toxiproxy like this:
listen: localhost:1521, upstream: db.mydomain.com:1521
OR
listen: localhost:1521, upstream: 1.2.3.4:1521

Then I am unable to connect and see error "ORA-12516: TNS: listener could not find available handler with matching protocol stack."

It appears that the Oracle database server is then initiating a disconnect and my JDBC application throws a connection exception.

Is there any debugging I can turn up on the proxy to see what else the server might be sending back?  Given that it still fails even if I point toxiproxy to one specific upstream IP I can only assume Oracle is sending something back over the proxy that toxi doesn't understand.

Henry
